### ISSUE TITLE
Variables inside list comprehensions are treated as locals.

### ISSUE THREAD

**Issue by lensvol** [wemake-services/wemake-python-styleguide#247]
# Bug report

## What's wrong

When rule about restricting the number of local variables is applied to function code it makes an error
of counting variables inside list comprehensions towards the total number of local variables. It then triggers `TooManyLocalsViolation`, which is, in that case, may seem unneeded.

Example:

```python
def func():
      a = 1
      l = [
          n * i  
          for n in range(42) 
          for i in range(n)
      ]

```

There are only 2 local variables here, but code will count 4.

## How is that should be

Variables inside comprehensions (both list and dictionary, possibly including generators) should not count toward the total number of local variables.

### flake8 information

Contents of `flake8 --bug-report`:

```
{
  "dependencies": [
    {
      "dependency": "setuptools",
      "version": "39.0.1"
    }
  ],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.6.5",
    "system": "Darwin"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "flake8-bandit",
      "version": "v1.0.2"
    },
    {
      "is_local": false,
      "plugin": "flake8-broken-line",
      "version": "0.1.0"
    },
    {
      "is_local": false,
      "plugin": "flake8-bugbear",
      "version": "18.8.0"
    },
    {
      "is_local": false,
      "plugin": "flake8-comprehensions",
      "version": "1.4.1"
    },
    {
      "is_local": false,
      "plugin": "flake8-debugger",
      "version": "3.1.0"
    },
    {
      "is_local": false,
      "plugin": "flake8-docstrings",
      "version": "1.3.0, pydocstyle: 2.1.1"
    },
    {
      "is_local": false,
      "plugin": "flake8-eradicate",
      "version": "0.1.1"
    },
    {
      "is_local": false,
      "plugin": "flake8-print",
      "version": "3.1.0"
    },
    {
      "is_local": false,
      "plugin": "flake8-pytest",
      "version": "1.3"
    },
    {
      "is_local": false,
      "plugin": "flake8-string-format",
      "version": "0.2.3"
    },
    {
      "is_local": false,
      "plugin": "flake8-type-annotations",
      "version": "0.1.0"
    },
    {
      "is_local": false,
      "plugin": "flake8_builtins",
      "version": "1.4.1"
    },
    {
      "is_local": false,
      "plugin": "flake8_coding",
      "version": "1.3.1"
    },
    {
      "is_local": false,
      "plugin": "flake8_commas",
      "version": "2.0.0"
    },
    {
      "is_local": false,
      "plugin": "flake8_isort",
      "version": "2.3"
    },
    {
      "is_local": false,
      "plugin": "flake8_pep3101",
      "version": "1.2.1"
    },
    {
      "is_local": false,
      "plugin": "flake8_quotes",
      "version": "1.0.0"
    },
    {
      "is_local": false,
      "plugin": "logging-format",
      "version": "0.5.0"
    },
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "naming",
      "version": "0.7.0"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.3.1"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "1.6.0"
    },
    {
      "is_local": false,
      "plugin": "wemake-python-styleguide",
      "version": "0.2.0"
    }
  ],
  "version": "3.5.0"
}
```

### pip information

Contents of `pip freeze`:

```
added-value==0.8
alabaster==0.7.12
appnope==0.1.0
atomicwrites==1.2.1
attrs==18.2.0
Babel==2.6.0
backcall==0.1.0
bandit==1.5.1
certifi==2018.8.24
chardet==3.0.4
coverage==4.5.1
decorator==4.3.0
doc8==0.8.0
docutils==0.14
eradicate==0.2.1
flake8==3.5.0
flake8-bandit==1.0.2
flake8-broken-line==0.1.0
flake8-bugbear==18.8.0
flake8-builtins==1.4.1
flake8-coding==1.3.1
flake8-commas==2.0.0
flake8-comprehensions==1.4.1
flake8-debugger==3.1.0
flake8-docstrings==1.3.0
flake8-eradicate==0.1.1
flake8-isort==2.5
flake8-logging-format==0.5.0
flake8-pep3101==1.2.1
flake8-polyfill==1.0.2
flake8-print==3.1.0
flake8-pytest==1.3
flake8-quotes==1.0.0
flake8-string-format==0.2.3
flake8-type-annotations==0.1.0
gitdb2==2.0.4
GitPython==2.1.11
idna==2.7
imagesize==1.1.0
ipdb==0.11
ipython==7.0.1
ipython-genutils==0.2.0
isort==4.3.4
jedi==0.13.1
Jinja2==2.10
m2r==0.2.0
MarkupSafe==1.0
mccabe==0.6.1
mistune==0.8.3
more-itertools==4.3.0
mypy==0.630
mypy-extensions==0.4.1
packaging==18.0
parso==0.3.1
pbr==4.2.0
pep8-naming==0.7.0
pexpect==4.6.0
pickleshare==0.7.5
Pillow==5.3.0
pluggy==0.7.1
pockets==0.7.1
prompt-toolkit==2.0.5
ptyprocess==0.6.0
py==1.6.0
pycodestyle==2.3.1
pydocstyle==2.1.1
pyflakes==1.6.0
Pygments==2.2.0
pyparsing==2.2.2
pytest==3.8.2
pytest-cov==2.6.0
pytest-flake8==1.0.2
pytest-isort==0.2.1
pytest-randomly==1.2.3
pytz==2018.5
PyYAML==3.13
requests==2.19.1
restructuredtext-lint==1.1.3
simplegeneric==0.8.1
six==1.11.0
smmap2==2.0.4
snowballstemmer==1.2.1
Sphinx==1.8.1
sphinx-autodoc-typehints==1.3.0
sphinx-readable-theme==1.3.0
sphinxcontrib-napoleon==0.7
sphinxcontrib-websupport==1.1.0
stevedore==1.29.0
testfixtures==6.3.0
tomlkit==0.4.4
traitlets==4.3.2
typed-ast==1.1.0
typing==3.6.6
typing-extensions==3.6.5
urllib3==1.23
wcwidth==0.1.7
-e git+git@github.com:lensvol/wemake-python-styleguide.git@1310c45462c6fe1b6756c5de075314564fd6d917#egg=wemake_python_styleguide
```
